cheatsheets_draft/GraphQL_Cheat_Sheet.md

```diff
@@ -0,0 +1,333 @@
+
+#### Query Limiting (Depth & Amount)
+
+In GraphQL each query has a depth (_e.g._ nested objects) and each object requested in a query can have an amount specified (_e.g._ 99999999 of an object). By default these can both be unlimited which may lead to a DoS. You should set limits on depth and amount to prevent DoS, but this usually requires a small custom implementation as it is not natively supported by GraphQL. See [this](https://www.apollographql.com/blog/securing-your-graphql-api-from-malicious-queries-16130a324a6b) and [this](https://www.howtographql.com/advanced/4-security/) page for more information about these attacks and how to add depth and amount limiting. Adding [pagination](https://graphql.org/learn/pagination/) can also help performance.
+
+APIs using graphql-java can utilize the built-in [MaxQueryDepthInstrumentation](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/MaxQueryDepthInstrumentation.java) for depth limiting. APIs using JavaScript can use [graphql-depth-limit](https://github.com/stems/graphql-depth-limit) to implement depth limiting and [graphql-input-number](https://github.com/joonhocho/graphql-input-number) to implement amount limiting.
+
+Here is an example of a GraphQL query with depth N:
+
+```javascript
+query evil {            # Depth: 0
+  album(id: 42) {       # Depth: 1
+    songs {             # Depth: 2
+      album {           # Depth: 3
+        ...             # Depth: ...
+        album {id: N}   # Depth: N
+      }
+    }
+  }
+}
+```
+
+Here is an example of a GraphQL query requesting 99999999 of an object:
+
+```javascript
+query {
+  author(id: "abc") {
+    posts(first: 99999999) {
+      title
+    }
+  }
+}
+```
+
+#### Timeouts
+
+Adding timeouts can be a simple way to limit how many resources any single request can consume. But timeouts are not always effective since they may not activate until a malicious query has already consumed excessive resources. Timeout requirements will differ by API and data fetching mechanism; there isn't one timeout value that will work across the board.
+
+At the application level, timeouts can be added for queries and resolver functions. This option is usually more effective since the query/resolution can be stopped once the timeout is reached. GraphQL does not natively support query timeouts so custom code is required. See [this blog post](https://medium.com/workflowgen/graphql-query-timeout-and-complexity-management-fab4d7315d8d) for more about using timeouts with GraphQL or the two examples below.
+
+***JavaScript Timeout Example***
+
+Code snippet from [this SO answer](https://stackoverflow.com/a/53277955/1200388):
+
+```javascript
+request.incrementResolverCount =  function () {
+    var runTime = Date.now() - startTime;
+    if (runTime > 10) {  // a timeout of 10 seconds
```

Comment:
Since you're using [`Date.now()`][1], you'll getting a timestamp in milliseconds, meaning that the difference between `Date.now()` and `startTime` should also be in milliseconds.

```suggestion
    var runTime = Date.now() - startTime;
    if (runTime > 10000) {  // a timeout of 10 seconds
```


[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now

Reply:
@bigshebang you almost DoSed the service :yum: